source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,195 @@
+  absl::flat_hash_map<std::string, MetricsPluginPtr> name_to_custom_metrics_plugin_map;
+  for (const envoy::config::core::v3::TypedExtensionConfig& config :
+       spec.metrics_plugin_configs()) {
+    absl::StatusOr<MetricsPluginPtr> metrics_plugin_or = LoadMetricsPlugin(config);
+    RELEASE_ASSERT(
+        metrics_plugin_or.ok(),
+        absl::StrCat(
+            "MetricsPlugin loading error should have been caught during input validation: ",
+            metrics_plugin_or.status().message()));
+    name_to_custom_metrics_plugin_map[config.name()] = std::move(metrics_plugin_or.value());
+  }
+  return name_to_custom_metrics_plugin_map;
+}
+
+/**
+ * Loads and initializes a StepController plugin requested in the session spec. Assumes
+ * the spec has already been validated; crashes the process otherwise.
+ *
+ * @param spec Adaptive load session spec proto that has already been validated.
+ *
+ * @return unique_ptr<StepController> Initialized StepController plugin.
+ */
+StepControllerPtr LoadStepControllerPluginFromSpec(const AdaptiveLoadSessionSpec& spec) {
+  absl::StatusOr<StepControllerPtr> step_controller_or =
+      LoadStepControllerPlugin(spec.step_controller_config(), spec.nighthawk_traffic_template());
+  RELEASE_ASSERT(
+      step_controller_or.ok(),
+      absl::StrCat(
+          "StepController plugin loading error should have been caught during input validation: ",
+          step_controller_or.status().message()));
+  return std::move(step_controller_or.value());
+}
+
+} // namespace
+
+AdaptiveLoadControllerImpl::AdaptiveLoadControllerImpl(
+    const NighthawkServiceClient& nighthawk_service_client,
+    const MetricsEvaluator& metrics_evaluator,
+    const AdaptiveLoadSessionSpecProtoHelper& session_spec_proto_helper,
+    Envoy::TimeSource& time_source)
+    : nighthawk_service_client_{nighthawk_service_client}, metrics_evaluator_{metrics_evaluator},
+      session_spec_proto_helper_{session_spec_proto_helper}, time_source_{time_source} {}
+
+absl::StatusOr<BenchmarkResult> AdaptiveLoadControllerImpl::PerformAndAnalyzeNighthawkBenchmark(
+    nighthawk::client::NighthawkService::StubInterface* nighthawk_service_stub,
+    const AdaptiveLoadSessionSpec& spec,
+    const absl::flat_hash_map<std::string, MetricsPluginPtr>& name_to_custom_plugin_map,
+    const StepController& step_controller, Envoy::ProtobufWkt::Duration duration) {
+  absl::StatusOr<nighthawk::client::CommandLineOptions> command_line_options_or =
+      step_controller.GetCurrentCommandLineOptions();
+  if (!command_line_options_or.ok()) {
+    ENVOY_LOG_MISC(error, "Error constructing Nighthawk input: {}: {}",
+                   command_line_options_or.status().code(),
+                   command_line_options_or.status().message());
+    return command_line_options_or.status();
+  }
+  nighthawk::client::CommandLineOptions command_line_options = command_line_options_or.value();
+  *command_line_options.mutable_duration() = std::move(duration);
```

Comment:
it still feels a little bit off to me that we're specializing the way we treat the test duration. maybe this is a good place to add a comment about why we need that?

Reply:
Added some comments for now -- see issue #539 for a potential refactoring.